dataservices/dataservices.go

```diff
@@ -0,0 +1,239 @@
+}
+
+//NewDataServiceClient returns an initialised client
+func NewDataServiceClient() *DataServiceClient {
+	log.Info(version.Long())
+
+	userClient := user.NewServicesClient()
+	userClient.Start()
+
+	return &DataServiceClient{
+		api: rest.NewApi(),
+		//TODO: from config
+		dataStore:        store.NewMongoStore("deviceData"),
+		validateToken:    user.NewAuthorizationMiddleware(userClient).ValidateToken,
+		attachPermissons: user.NewMetadataMiddleware(userClient).GetPermissons,
+		resolveGroupID:   user.NewMetadataMiddleware(userClient).GetGroupID,
+	}
+
+}
+
+//Run will run the service
+func (client *DataServiceClient) Run(URL string) error {
+	client.api.Use(rest.DefaultDevStack...)
+	client.api.Use(&rest.GzipMiddleware{})
+
+	router, err := rest.MakeRouter(
+		rest.Get("/version", client.GetVersion),
+		rest.Get("/data/:userid/:datumid", client.validateToken(client.resolveGroupID((client.GetData)))),
+		rest.Post("/dataset/:userid", client.validateToken(client.attachPermissons(client.resolveGroupID(client.PostDataset)))),
+		rest.Get("/dataset/:userid", client.validateToken(client.attachPermissons(client.resolveGroupID(client.GetDataset)))),
+		rest.Post("/blob/:userid", client.validateToken(client.attachPermissons(client.resolveGroupID(client.PostBlob)))),
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+	client.api.SetApp(router)
+
+	return http.ListenAndServe(URL, client.api.MakeHandler())
+}
+
+//checkPermisson will check that we have the expected permisson
+func checkPermisson(r *rest.Request, expected user.Permission) bool {
+	//TODO: fill in the details
+	if permissions := r.Env[user.PERMISSIONS].(*user.UsersPermissions); permissions != nil {
+		return true
+	}
+	return false
+}
+
+//GetVersion will return the current API version
+func (client *DataServiceClient) GetVersion(w rest.ResponseWriter, r *rest.Request) {
+	w.WriteJson(version.Long())
+	return
+}
+
+//PostDataset will process a posted dataset for the requested user if permissons are sufficient
+func (client *DataServiceClient) PostDataset(w rest.ResponseWriter, r *rest.Request) {
+
+	userid := r.PathParam(useridParamName)
+
+	log.AddTrace(userid)
+
+	if checkPermisson(r, user.Permission{}) {
+
+		groupID := r.Env[user.GROUPID]
+
+		if r.ContentLength == 0 || groupID == "" {
+			rest.Error(w, missingDataError, http.StatusBadRequest)
+			return
+		}
+
+		var dataSet data.Dataset
+		var processedDataset struct {
+			Dataset []interface{} `json:"Dataset"`
+			Errors  string        `json:"Errors"`
+		}
+
+		err := r.DecodeJsonPayload(&dataSet)
+
+		if err != nil {
+			rest.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		platformData, err := data.NewTypeBuilder(map[string]interface{}{data.UserIDField: userid, data.GroupIDField: groupID}).BuildFromDataSet(dataSet)
+		processedDataset.Dataset = platformData
+		processedDataset.Errors = err.Error()
+
+		if err.Error() != "" {
+			w.WriteHeader(http.StatusBadRequest)
+			w.WriteJson(&processedDataset)
```

Comment:
We shouldn't be returning any data here, I don't think  Certainly not on an error.